### PR TITLE
Frontend refactor

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This is another frontend refactoring. This time it extracts common code between Frontend v1 and scheduler, specifically the queue. Queue has been moved into its own package (`pkg/scheduler/queue`, but should perhaps be moved to `pkg/util/queue`), and now has a type `*RequestQueue`. This is used by both frontendv1 and scheduler. Some benchmarks from original frontend package were moved here as well.

The only logical thing that has changed is discarding of the expired requests -- previously this was done in `GetNextRequestForQuerier` method (in both scheduler and frontend), but since queue now doesn't deal with request in any way (request is now basically `interface{}`), it has been moved into scheduler/frontend instead.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/frontend/v1/queue_test.go [hunk 1]
```diff
@@ -39,194 +49,118 @@ func TestDequeuesExpiredRequests(t *testing.T) {
...
-	require.Nil(t, err)
-	require.NotNil(t, req)
-
-	// ensure either one or two queues are fully drained, depending on which was requested first
-	_, ok := f.queues.userQueues[userID]
-	if ok {
-		// if the second user's queue was chosen for the last request,
-		// the first queue should still contain 4 (expired) requests.
-		require.Equal(t, 4, len(f.queues.getOrAddQueue(userID, 0)))
+	// Replay entire conversation in one second.
```
Comment (pracucci) [inline pkg/frontend/v1/queue_test.go:73]: I find this comment unclear. Is there a better way to phrase it?

File: pkg/frontend/v1/queue_test.go [hunk 2]
```diff
@@ -39,194 +49,118 @@ func TestDequeuesExpiredRequests(t *testing.T) {
...
 	}
 
-	ctx := context.Background()
-	idx := -1
-	for i := 0; i < 100; i++ {
-		req, nidx, err := f.getNextRequestForQuerier(ctx, idx, "")
-		require.NoError(t, err)
-		require.NotNil(t, req)
-		idx = nidx
+	// Replay entire conversation.
```
Comment (pracucci) [inline pkg/frontend/v1/queue_test.go:110]: Same here. I would clarify this comment.

**Review by pracucci — APPROVED** [review]
Wow, good job! It took me a while to compare with the current code in `master` but I haven't noticed any issue! I just left few non blocking comments. Thanks!

File: pkg/frontend/v1/queue_test.go (hunk 1 above)
Comment (pstibrany) [reply pkg/frontend/v1/queue_test.go]: I've tried:

```
	// Calling Process will only return when client disconnects or context is finished.
	// We use context timeout to stop Process call.
```

**Comment by pstibrany** [comment]
Thanks for your review @pracucci!

File: pkg/frontend/v1/queue_test.go (hunk 1 above)
Comment (pracucci) [reply pkg/frontend/v1/queue_test.go]: Way more clear, thanks!

**Comment by pracucci** [comment]
Thanks @pstibrany ! Let's ship this ;)

**Pull request merged**